### ISSUE TITLE
CH 21.6.4: Number of arguments for function bitShiftRight doesn't match

### ISSUE THREAD

**Issue by vmihailenco** [ClickHouse/ClickHouse#25233]
```sql
CREATE TABLE test (
  id UInt64
)
ENGINE = MergeTree()
PARTITION BY intDiv(bitShiftRight(id, 1), 1000)
ORDER BY id;

INSERT INTO test VALUES (1);

SELECT * FROM test WHERE id = 1;
```

produces

```
Received exception from server (version 21.6.4):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Number of arguments for function bitShiftRight doesn't match: passed 1, should be 2.
```

Same code works on 21.5 and before.

**Comment by kssenii** [comment]
hm, this example on 21.6.4.26 works:

```
kssenii.ru-central1.internal :) select version();

SELECT version()

Query id: 5fee9907-be50-4df1-ab37-92d2cfa68612

┌─version()─┐
│ 21.6.4.26 │
└───────────┘

1 rows in set. Elapsed: 0.002 sec. 

kssenii.ru-central1.internal :) CREATE TABLE test (
:-]   id UInt64
:-] )
:-] ENGINE = MergeTree()
:-] PARTITION BY intDiv(bitShiftRight(id, 1), 1000)
:-] ORDER BY id;

CREATE TABLE test
(
    `id` UInt64
)
ENGINE = MergeTree
PARTITION BY intDiv(bitShiftRight(id, 1), 1000)
ORDER BY id

Query id: d2bcfae6-2ea3-4940-b807-afb5918e0a29

Ok.

0 rows in set. Elapsed: 0.014 sec. 

kssenii.ru-central1.internal :) INSERT INTO test VALUES (1);

INSERT INTO test VALUES

Query id: 0816de41-caba-4242-8135-503acb9505ad

Ok.

1 rows in set. Elapsed: 0.002 sec. 

kssenii.ru-central1.internal :) SELECT * FROM test WHERE id = 1;

SELECT *
FROM test
WHERE id = 1

Query id: a9d825fa-1f78-4474-b130-9e254549b252

┌─id─┐
│  1 │
└────┘

1 rows in set. Elapsed: 0.002 sec. 

```
In version 21.7 also works:
```
kssenii.ru-central1.internal :) select version();

SELECT version()

Query id: d1716551-f4e7-4f55-b9b7-dc5be57b8501

┌─version()─┐
│ 21.7.1.1  │
└───────────┘

1 rows in set. Elapsed: 0.004 sec. 

kssenii.ru-central1.internal :) CREATE TABLE test (
:-]   id UInt64
:-] )
:-] ENGINE = MergeTree()
:-] PARTITION BY intDiv(bitShiftRight(id, 1), 1000)
:-] ORDER BY id;

CREATE TABLE test
(
    `id` UInt64
)
ENGINE = MergeTree
PARTITION BY intDiv(bitShiftRight(id, 1), 1000)
ORDER BY id

Query id: 9b427c6f-fa94-48d3-95c2-9e08ff0adf34

Ok.

0 rows in set. Elapsed: 0.012 sec. 

kssenii.ru-central1.internal :) INSERT INTO test VALUES (1);

INSERT INTO test VALUES

Query id: dd945a32-2d4d-44c3-868c-0324aecfa495

Ok.

1 rows in set. Elapsed: 0.006 sec. 

kssenii.ru-central1.internal :) SELECT * FROM test WHERE id = 1;

SELECT *
FROM test
WHERE id = 1

Query id: f7ec378b-757a-4c70-a87c-b59fa77c62b2

┌─id─┐
│  1 │
└────┘

1 rows in set. Elapsed: 0.007 sec. 

```

I also checked another case - creating the same table on version 21.3 and then upgrading server to the latest version and it still works.

**Comment by vmihailenco** [comment]
Indeed - very strange :/ I will try to find a better way to reproduce this. Here is the backtrace from the log

```
2021.06.12 12:08:46.043122 [ 302590 ] {8e31e720-e3e3-46dd-9d9f-b798fb7e090d} <Error> TCPHandler: Code: 42, e.displayText() = DB::Exception: Number of arguments for function bitShiftRight doesn't match: passed 1, should be 2, Stack trace:

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8b5e67a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) @ 0xa7760d3 in /usr/bin/clickhouse
2. DB::IFunctionOverloadResolver::checkNumberOfArguments(unsigned long) const @ 0xefccb0f in /usr/bin/clickhouse
3. DB::IFunctionOverloadResolver::getReturnTypeWithoutLowCardinality(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&) const @ 0xefcd1f5 in /usr/bin/clickhouse
4. DB::IFunctionOverloadResolver::getReturnType(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&) const @ 0xefcce7d in /usr/bin/clickhouse
5. ? @ 0x100af1eb in /usr/bin/clickhouse
6. DB::KeyCondition::canConstantBeWrappedByFunctions(std::__1::shared_ptr<DB::IAST> const&, unsigned long&, std::__1::shared_ptr<DB::IDataType const>&, DB::Field&, std::__1::shared_ptr<DB::IDataType const>&) @ 0x100afec7 in /usr/bin/clickhouse
7. DB::KeyCondition::tryParseAtomFromAST(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::Block&, DB::KeyCondition::RPNElement&) @ 0x100ad051 in /usr/bin/clickhouse
8. DB::KeyCondition::traverseAST(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::Block&) @ 0x100ac0f8 in /usr/bin/clickhouse
9. DB::KeyCondition::traverseAST(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::Block&) @ 0x100ac3d5 in /usr/bin/clickhouse
10. DB::KeyCondition::traverseAST(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::Block&) @ 0x100ac3d5 in /usr/bin/clickhouse
11. DB::KeyCondition::KeyCondition(DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context>, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::ExpressionActions> const&, bool, bool) @ 0x100ab65b in /usr/bin/clickhouse
12. DB::PartitionPruner::PartitionPruner(std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context>, bool) @ 0x100d4164 in /usr/bin/clickhouse
13. DB::MergeTreeDataSelectExecutor::readFromParts(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > >, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context>, unsigned long, unsigned int, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, long> > > const*, DB::MergeTreeDataSelectCache*) const @ 0x101a6728 in /usr/bin/clickhouse
14. DB::MergeTreeDataSelectExecutor::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context>, unsigned long, unsigned int, DB::QueryProcessingStage::Enum, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, long> > > const*) const @ 0x101a1c6b in /usr/bin/clickhouse
15. DB::StorageMergeTree::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context>, DB::QueryProcessingStage::Enum, unsigned long, unsigned int) @ 0xfe5a36d in /usr/bin/clickhouse
16. DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&) @ 0xf8a1eaf in /usr/bin/clickhouse
17. DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::shared_ptr<DB::IBlockInputStream> const&, std::__1::optional<DB::Pipe>) @ 0xf899c03 in /usr/bin/clickhouse
18. DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0xf898cc0 in /usr/bin/clickhouse
19. DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) @ 0xfa0875e in /usr/bin/clickhouse
20. DB::InterpreterSelectWithUnionQuery::execute() @ 0xfa09831 in /usr/bin/clickhouse
21. ? @ 0xfbc4221 in /usr/bin/clickhouse
22. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, bool) @ 0xfbc28a3 in /usr/bin/clickhouse
23. DB::TCPHandler::runImpl() @ 0x103e8d92 in /usr/bin/clickhouse
24. DB::TCPHandler::run() @ 0x103fbc99 in /usr/bin/clickhouse
25. Poco::Net::TCPServerConnection::start() @ 0x129f93af in /usr/bin/clickhouse
26. Poco::Net::TCPServerDispatcher::run() @ 0x129fae3a in /usr/bin/clickhouse
27. Poco::PooledThread::run() @ 0x12b34e39 in /usr/bin/clickhouse
28. Poco::ThreadImpl::runnableEntry(void*) @ 0x12b30e2a in /usr/bin/clickhouse
29. start_thread @ 0x9609 in /lib/x86_64-linux-gnu/libpthread-2.31.so
30. /build/glibc-eX1tMB/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97: clone @ 0x122293 in /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.31.so
```

Is this enough?

**Comment by vmihailenco** [comment]
PS Hm, I can reproduce this with queries from the first message, but only **after** I create a bunch of other tables. But if I create the test table first, then I get no errors...

**Comment by vmihailenco** [comment]
PPS Furthermore, if I create a bunch of tables and then restart Clickhouse, then all queries are working as expected again. I guess I can live with that :)

**Comment by kssenii** [comment]
> PS Hm, I can reproduce this with queries from the first message, but only after I create a bunch of other tables. But if I create the test table first, then I get no errors...

hm, interesting...

> I guess I can live with that :)

Anyway, this needs to be investigated.

**Comment by kitaisreal** [comment]
It is related to JIT compilation, https://github.com/ClickHouse/ClickHouse/pull/24273 should fix that, but I assumed that it was merged before 21.6 checkouted from master.
For fix you can disable expressions compilation. Set compile_expressions to 0.

**Comment by kitaisreal** [comment]
Fixed with backport https://github.com/ClickHouse/ClickHouse/pull/25240.
